drivers/sdhc/sdhc_esp32.c

```diff
@@ -0,0 +1,1544 @@
+/*
+ * Copyright (c) 2024 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_sdhc
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sdhc.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+#include <zephyr/drivers/pinctrl.h>
+
+/* ESP32 includes */
+#include <esp_clk_tree.h>
+#include <hal/sdmmc_ll.h>
+#include <esp_intr_alloc.h>
+#include <esp_timer.h>
+#include <hal/gpio_hal.h>
+#include <hal/rtc_io_hal.h>
+#include <soc/sdmmc_reg.h>
+#include <esp_memory_utils.h>
+
+#include "sdhc_esp32.h"
+
+LOG_MODULE_REGISTER(sdhc, CONFIG_SDHC_LOG_LEVEL);
+
+#define SDMMC_SLOT_WIDTH_DEFAULT 1
+
+#define SDMMC_HOST_CLOCK_UPDATE_CMD_TIMEOUT_US 1000 * 1000
+#define SDMMC_HOST_RESET_TIMEOUT_US            5000 * 1000
+#define SDMMC_HOST_START_CMD_TIMEOUT_US        1000 * 1000
+#define SDMMC_HOST_WAIT_EVENT_TIMEOUT_US       1000 * 1000
+
+#define SDMMC_EVENT_QUEUE_LENGTH 32
+
+#define SDMMC_TIMEOUT_MAX 0xFFFFFFFF
+
+/* Number of DMA descriptors used for transfer.
+ * Increasing this value above 4 doesn't improve performance for the usual case
+ * of SD memory cards (most data transfers are multiples of 512 bytes).
+ */
+#define SDMMC_DMA_DESC_CNT 4
+
+struct sdhc_esp32_config {
+
+	int index;
+	int slot;
+	const sdmmc_dev_t *sdio_hw;
+	const struct pinctrl_dev_config *pcfg;
+	const struct gpio_dt_spec pwr_gpio;
+	int irq_source;
+	uint8_t bus_width_cfg;
+
+	/* Bindings properties */
+	uint32_t max_current_330;
+	uint32_t max_current_180;
+	uint32_t power_delay_ms;
+	uint32_t min_bus_freq;
+	uint32_t max_bus_freq;
+};
+
+struct sdhc_esp32_data {
+
+	uint8_t bus_width;  /* Bus width used by the slot (can change during execution) */
+	uint32_t bus_clock; /* Value in Hz. ESP-IDF functions use kHz instead */
+
+	enum sdhc_power power_mode;
+	enum sdhc_timing_mode timing;
+
+	struct sdhc_host_props props;
+};
+
+/**********************************************************************
+ * SDHC global symbols
```

Comment:
Should these global variables be members of the device data struct? It seems so, since I imagine each instance of the ESP32 SDHC controller will need a separate request mutex for example